### PR TITLE
Feature/more profile screens

### PR DESCRIPTION
**Pull request by hedvigoscar**
The rest of the feature, ya know

### CODE REVIEW

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/charity/CharityFragment.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        loadData()
+    }
+
+    private fun loadData() {
+        profileViewModel.data.observe(this, Observer { profileData ->
+            loadingSpinner.remove()
+
+            profileData?.let { data ->
```
Comment (FredrikAreschoug) [inline owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/charity/CharityFragment.kt]: Can't we already here check `profileData?.cashback()` ?

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/feedback/FeedbackFragment.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        super.onViewCreated(view, savedInstanceState)
+
+        (requireActivity() as AppCompatActivity).setSupportActionBar(toolbar)
+
+        collapsingToolbar.title = resources.getString(R.string.PROFILE_FEEDBACK_TITLE)
+        collapsingToolbar.setExpandedTitleTypeface(requireContext().compatFont(R.font.circular_bold))
+        collapsingToolbar.setCollapsedTitleTypeface(requireContext().compatFont(R.font.circular_bold))
+        toolbar.setNavigationIcon(R.drawable.ic_back)
+        toolbar.setNavigationOnClickListener {
+            val intent = Intent("profileNavigation")
```
Comment (FredrikAreschoug) [inline owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/feedback/FeedbackFragment.kt]: profileNavigation as a constant?

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myhome/MyHomeFragment.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        (requireActivity() as AppCompatActivity).setSupportActionBar(toolbar)
+
+        collapsingToolbar.title = resources.getString(R.string.PROFILE_MY_HOME_TITLE)
+        collapsingToolbar.setExpandedTitleTypeface(requireContext().compatFont(R.font.circular_bold))
+        collapsingToolbar.setCollapsedTitleTypeface(requireContext().compatFont(R.font.circular_bold))
+        toolbar.setNavigationIcon(R.drawable.ic_back)
+        toolbar.setNavigationOnClickListener {
```
Comment (FredrikAreschoug) [inline owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myhome/MyHomeFragment.kt]: Setting up `collapsingToolbar` and `toolbar` is basacally the same in feedbackfragment should we crate an extension?

File: owldroid/src/main/res/layout/profile_menu_row.xml [hunk 4]
```diff
@@ -47,6 +49,7 @@
     <ImageView
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
+        android:contentDescription="@null"
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/res/layout/profile_menu_row.xml]: Non-descriptive image :)

File: owldroid/src/main/res/layout/fragment_payment.xml [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+                android:layout_marginTop="24dp"
+                android:visibility="gone"
+                app:layout_constraintEnd_toEndOf="parent"
+                app:layout_constraintStart_toStartOf="parent"
+                app:layout_constraintTop_toTopOf="parent"
+                tools:visibility="visible">
+
+                <ImageView
+                    android:id="@id/priceSphere"
+                    android:layout_width="160dp"
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/res/layout/fragment_payment.xml]: 👍

File: owldroid/src/main/res/values/styles.xml [hunk 6]
```diff
@@ -89,4 +87,49 @@
         <item name="android:actionMenuTextAppearance">@style/AppTheme.MenuItem</item>
     </style>
 
+    <style name="DialogTitle">
+        <item name="android:textSize">16sp</item>
+        <item name="android:fontFamily">@font/circular_bold</item>
+        <item name="android:textColor">@color/off_black_dark</item>
+    </style>
+
+    <style name="DialogParagraph">
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/res/values/styles.xml]: 👍

File: owldroid/src/main/res/layout/loading_spinner.xml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ProgressBar xmlns:android="http://schemas.android.com/apk/res/android"
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/res/layout/loading_spinner.xml]: :+1:

File: owldroid/src/main/res/anim/slide_in_left.xml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:duration="300"
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/res/anim/slide_in_left.xml]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/util/Validation.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+import android.support.annotation.StringRes
+import com.hedvig.android.owldroid.R
+
+object Regexes {
+    val emailRegex = Regex("^\\S+@\\S+\$")
+    val phoneNumberRegex = Regex("([+]*[0-9]+[+. -]*)")
+}
+
+fun validateEmail(email: CharSequence): ValidationResult {
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/util/Validation.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/payment/PaymentFragment.kt [hunk 10]
```diff
@@ -6,29 +6,35 @@ import android.content.Context
...
+import com.hedvig.android.owldroid.util.extensions.show
+import com.hedvig.android.owldroid.util.interpolateTextKey
 import dagger.android.support.AndroidSupportInjection
-import kotlinx.android.synthetic.main.activity_payment.*
+import kotlinx.android.synthetic.main.app_bar.*
+import kotlinx.android.synthetic.main.fragment_payment.*
+import java.util.Calendar
 import javax.inject.Inject
 
+const val BILLING_DAY = 27
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/payment/PaymentFragment.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/payment/PaymentFragment.kt [hunk 11]
```diff
@@ -43,47 +49,112 @@ class PaymentFragment : Fragment() {
...
-            profile_payment_change_bank_account.background.compatSetTint(ContextCompat.getColor(context!!, R.color.dark_purple))
-
-            profile_payment_price_sphere.drawable.compatSetTint(ContextCompat.getColor(context!!, R.color.green))
-            val monthlyCost = profileData!!.insurance().monthlyCost()?.toString()
+            loadingSpinner.remove()
+            sphereContainer.show()
+
+            val monthlyCost = profileData?.insurance()?.monthlyCost()?.toString()
             val amountPartOne = SpannableString("$monthlyCost\n")
             val perMonthLabel = "kr/mån"
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/payment/PaymentFragment.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myinfo/ValidationDialog.kt [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+class ValidationDialog : DialogFragment() {
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
+        inflater.inflate(R.layout.dialog_validation, container, false)
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        arguments?.let { args ->
+            dialogTitle.text = resources.getString(args.getInt("title"))
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myinfo/ValidationDialog.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myinfo/MyInfoFragment.kt [hunk 13]
```diff
@@ -82,33 +87,124 @@ class MyInfoFragment : Fragment() {
...
                 activity?.invalidateOptionsMenu()
             })
         })
     }
+
+    private fun setupEmailInput(prefilledEmail: String) {
+        emailInput.setText(prefilledEmail)
+
+        emailInput.onChange { value ->
+            profileViewModel.emailChanged(value)
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myinfo/MyInfoFragment.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/marketing/StoryFragment.kt [hunk 14]
```diff
@@ -43,17 +48,18 @@ class StoryFragment : Fragment() {
...
 
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-        val position = arguments?.getInt(POSITION_KEY) ?: throw Exception("No position provided")
+        val position = arguments?.getInt(POSITION_KEY) ?: throw Error("No position provided")
 
         val view = inflater.inflate(R.layout.page_marketing_story, container, false) as LinearLayout
-        val story = marketingStoriesViewModel.marketingStories.value?.get(position)
-                ?: throw Exception("No data")
+
+        val story = marketingStoriesViewModel.marketingStories.value?.get(position) ?: return View(context)
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/marketing/StoryFragment.kt]: Add a fixme

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/ProfileMenuRow.kt [hunk 15]
```diff
@@ -50,16 +53,24 @@ class ProfileMenuRow : ConstraintLayout {
...
 
     private fun setupDynamicContent() {
         val attributes = context.obtainStyledAttributes(attributeSet, R.styleable.ProfileMenuRow, defStyle, 0)
 
         icon = attributes.getDrawable(R.styleable.ProfileMenuRow_iconImage)
         name = attributes.getText(R.styleable.ProfileMenuRow_name)
-        description = attributes.getText(R.styleable.ProfileMenuRow_description)
+
+        val description = attributes.getText(R.styleable.ProfileMenuRow_description)
+        if (description == null) {
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/ProfileMenuRow.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/charity/CashbackOptionViewHolder.kt [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package com.hedvig.android.owldroid.ui.profile.charity
+
+import android.support.v7.widget.RecyclerView
+import android.view.View
+import android.widget.Button
+import android.widget.TextView
+import kotlinx.android.synthetic.main.cashback_option.view.*
+
+class CashbackOptionViewHolder(view: View) : RecyclerView.ViewHolder(view) {
```
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/charity/CashbackOptionViewHolder.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/charity/CharityFragment.kt (hunk 1 above)
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/charity/CharityFragment.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/feedback/FeedbackFragment.kt (hunk 2 above)
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/feedback/FeedbackFragment.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myhome/MyHomeFragment.kt (hunk 3 above)
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myhome/MyHomeFragment.kt]: :+1:

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myinfo/MyInfoFragment.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (hedvigoscar) [reply owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/myinfo/MyInfoFragment.kt]: Making this function generic for the case might just complicate the code more than we gain. Lets discuss later!

File: owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/referral/ReferralFragment.kt [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    override fun onStop() {
+        super.onStop()
+        buttonAnimator?.removeAllListeners()
+        buttonAnimator?.cancel()
+    }
+
+    private fun trackReferralSend() {
+        val bundle = Bundle()
+        profileViewModel.remoteConfigData.value?.let { rcd ->
+            bundle.putInt("incentive", rcd.referralsIncentiveAmount)
```
Comment (FredrikAreschoug) [inline owldroid/src/main/java/com/hedvig/android/owldroid/ui/profile/referral/ReferralFragment.kt]: "incentive" -> const

File: owldroid/src/main/res/layout/dialog_change_home_info.xml [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content">
+
+    <android.support.constraint.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:background="@color/off_white_light"
+        android:minWidth="280dp"
```
Comment (FredrikAreschoug) [inline owldroid/src/main/res/layout/dialog_change_home_info.xml]: move to dimen

File: owldroid/src/main/res/layout/fragment_referral.xml [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+
+    <android.support.v4.widget.NestedScrollView
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:background="@color/off_white"
+        app:layout_behavior="@string/appbar_scrolling_view_behavior">
+
+        <android.support.constraint.ConstraintLayout
+            android:layout_width="match_parent"
+            android:layout_marginBottom="60dp"
```
Comment (FredrikAreschoug) [inline owldroid/src/main/res/layout/fragment_referral.xml]: move to dimen

Comment (hedvigoscar) [reply owldroid/src/main/res/layout/fragment_referral.xml]: It's actually not super clear if we even want this marginBottom. It's related to a very strange bug that we can possibly evaluate in the office tomorrow

Comment (FredrikAreschoug) [reply owldroid/src/main/res/layout/fragment_referral.xml]: :+1:

**Pull request merged**